azurelinuxagent/common/utils/cryptutil.py

```diff
@@ -79,18 +86,50 @@ def decrypt_p7m(self, p7m_file, trans_prv_file, trans_cert_file, pem_file):
         elif not os.path.exists(trans_prv_file):
             raise IOError(errno.ENOENT, "File not found", trans_prv_file)
         else:
-            cmd = ("{0} cms -decrypt -in {1} -inkey {2} -recip {3} "
-                   "| {4} pkcs12 -nodes -password pass: -out {5}"
-                   "").format(self.openssl_cmd, p7m_file, trans_prv_file,
-                              trans_cert_file, self.openssl_cmd, pem_file)
-            shellutil.run(cmd)
-            rc = shellutil.run(cmd) # pylint: disable=C0103
-            if rc != 0:
-                logger.error("Failed to decrypt {0}".format(p7m_file))
+            first_cmd = [self.openssl_cmd, "cms", "-decrypt", "-in", p7m_file, "-inkey",
+                trans_prv_file, "-recip", trans_cert_file]
+            second_cmd = [self.openssl_cmd, "pkcs12", "-nodes", "-password", "pass:", 
+                "-out", pem_file]
+
+            first_proc = subprocess.Popen(first_cmd, stdout=subprocess.PIPE)
+
+            second_proc = subprocess.Popen(second_cmd, stdin=first_proc.stdout, stdout=subprocess.PIPE)
+            first_proc.stdout.close()  # see https://docs.python.org/2/library/subprocess.html#replacing-shell-pipeline
+            stdout, stderr = second_proc.communicate()
+            first_proc.poll()  # ensure that returncode is initialized
```

Comment:
`poll` doesn't wait for the process to terminate, whereas `wait` does (https://docs.python.org/2/library/subprocess.html#subprocess.Popen.poll). Is this intended?

Reply:
Either would work here, as first_proc will be completed by the time `second_proc.communicate`  returns. I thought that `poll` was more clear about what was happening; i.e. that we aren't waiting for `first_proc` on that line, it has already completed.

Reply:
@kevinclark19a we need to remove the check on the return code of the first command (the shell simply uses the return value of the last command in the pipe (unless one uses set -o pipefail))